
#Exercise
A customer has a site with a combined heat and power (CHP) plant and three gas boilers.
A CHP plant is a plant that takes natural gas as input and generates electricity and heat.
What can you say about the on-site electricity and heat generation? Gas consumption? And
the different units?

---



In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

In [2]:
!pwd

/content


## Data Set Overview

In [3]:
# Read the csv file
df = pd.read_csv('data_site.csv')

In [4]:
# Display the head of the dataset using the head() function
df.head()

From Timestamp  CHP Electricity Generated Day Total (kWh)  \
0  01/10/2022 00:00                                          0   
1  01/10/2022 00:30                                          0   
2  01/10/2022 01:00                                          0   
3  01/10/2022 01:30                                          0   
4  01/10/2022 02:00                                          0   

   CHP Electricity Generated Night Total (kWh)  \
0                                          540   
1                                          530   
2                                          480   
3                                          510   
4                                          510   

   Boiler 1 Heat Energy Total (MWh)  Boiler 2 Heat Energy Total (MWh)  \
0                               0.0                               0.1   
1                               0.0                               0.1   
2                               0.0                               0.3   
3                               0.0                               0.2   
4                               0.0                               0.1   

   Boiler 3 Heat Energy Total (MWh)  CHP Heat Energy Total (MWh)  \
0                               0.0                          0.5   
1                               0.0                          0.4   
2                               0.0                          0.4   
3                               0.0                          0.4   
4                               0.0                          0.5   

   Boiler 1 Gas Corrected Volume (Sm3)  Boiler 2 Gas Corrected Volume (Sm3)  \
0                                    0                                   19   
1                                    0                                   19   
2                                    0                                   25   
3                                    0                                   30   
4                                    0                                   15   

   Boiler 3 Gas Corrected Volume (Sm3)  \
0                                    0   
1                                    0   
2                                    0   
3                                    0   
4                                    0   

   CHP 2G Natural Gas Corrected Volume Total (Sm3)  
0                                              131  
1                                              129  
2                                              118  
3                                              122  
4                                              123

In [5]:
# Display the bottom 5 rows from the dataset using the tail() function
df.tail()

From Timestamp  CHP Electricity Generated Day Total (kWh)  \
1483  31/10/2022 21:30                                        550   
1484  31/10/2022 22:00                                        540   
1485  31/10/2022 22:30                                        550   
1486  31/10/2022 23:00                                        540   
1487  31/10/2022 23:30                                        540   

      CHP Electricity Generated Night Total (kWh)  \
1483                                            0   
1484                                            0   
1485                                            0   
1486                                            0   
1487                                            0   

      Boiler 1 Heat Energy Total (MWh)  Boiler 2 Heat Energy Total (MWh)  \
1483                               0.0                               0.0   
1484                               0.0                               0.0   
1485                               0.0                               0.0   
1486                               0.0                               0.0   
1487                               0.0                               0.0   

      Boiler 3 Heat Energy Total (MWh)  CHP Heat Energy Total (MWh)  \
1483                               0.3                          0.5   
1484                               0.2                          0.5   
1485                               0.3                          0.5   
1486                               0.3                          0.5   
1487                               0.3                          0.4   

      Boiler 1 Gas Corrected Volume (Sm3)  \
1483                                    0   
1484                                    0   
1485                                    0   
1486                                    0   
1487                                    0   

      Boiler 2 Gas Corrected Volume (Sm3)  \
1483                                    0   
1484                                    0   
1485                                    0   
1486                                    0   
1487                                    0   

      Boiler 3 Gas Corrected Volume (Sm3)  \
1483                                   28   
1484                                   35   
1485                                   40   
1486                                   35   
1487                                   34   

      CHP 2G Natural Gas Corrected Volume Total (Sm3)  
1483                                              133  
1484                                              134  
1485                                              134  
1486                                              134  
1487                                              132

### Observations:

Column *CHP Heat Energy Total (MWh)* values are in distinct units from CHP Electricity. The reading interval is every 30 minutes.


- Formula to convert units:
$$ \begin{equation} kWh = MWh \times 1000 \end{equation}$$

In [ ]:
# Converting the data of the data set to constant units (kWh)
df.iloc[:,3:7] *= 1000

In [ ]:
# Renaming all the columns for simplicity and uniform format
df.columns = ['Datetime','CHP_Electricity_Day(kWh)', 'CHP_Electricity_Night(kWh)', 'Boiler_1_Heat(kWh)', 'Boiler_2_Heat(kWh)', 'Boiler_3_Heat(kWh)', 'CHP_Total_Energy(kWh)', 'Boiler_1_Volume(Sm3)', 'Boiler_2_Volume(Sm3)','Boiler_3_Volume(Sm3)','CHP_2G_Volume(Sm3)']

In [ ]:
df.head()

Datetime  CHP_Electricity_Day(kWh)  CHP_Electricity_Night(kWh)  \
0  01/10/2022 00:00                         0                         540   
1  01/10/2022 00:30                         0                         530   
2  01/10/2022 01:00                         0                         480   
3  01/10/2022 01:30                         0                         510   
4  01/10/2022 02:00                         0                         510   

   Boiler_1_Heat(kWh)  Boiler_2_Heat(kWh)  Boiler_3_Heat(kWh)  \
0                 0.0               100.0                 0.0   
1                 0.0               100.0                 0.0   
2                 0.0               300.0                 0.0   
3                 0.0               200.0                 0.0   
4                 0.0               100.0                 0.0   

   CHP_Total_Energy(kWh)  Boiler_1_Volume(Sm3)  Boiler_2_Volume(Sm3)  \
0                  500.0                     0                    19   
1                  400.0                     0                    19   
2                  400.0                     0                    25   
3                  400.0                     0                    30   
4                  500.0                     0                    15   

   Boiler_3_Volume(Sm3)  CHP_2G_Volume(Sm3)  
0                     0                 131  
1                     0                 129  
2                     0                 118  
3                     0                 122  
4                     0                 123

## Feature Data Types

In [ ]:
# Summary of the data set information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1488 entries, 0 to 1487
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Datetime                    1488 non-null   object 
 1   CHP_Electricity_Day(kWh)    1488 non-null   int64  
 2   CHP_Electricity_Night(kWh)  1488 non-null   int64  
 3   Boiler_1_Heat(kWh)          1488 non-null   float64
 4   Boiler_2_Heat(kWh)          1488 non-null   float64
 5   Boiler_3_Heat(kWh)          1488 non-null   float64
 6   CHP_Total_Energy(kWh)       1488 non-null   float64
 7   Boiler_1_Volume(Sm3)        1488 non-null   int64  
 8   Boiler_2_Volume(Sm3)        1488 non-null   int64  
 9   Boiler_3_Volume(Sm3)        1488 non-null   int64  
 10  CHP_2G_Volume(Sm3)          1488 non-null   int64  
dtypes: float64(4), int64(6), object(1)
memory usage: 128.0+ KB


The provided data consists of over 1400 observations contained in 11 columns variables. 10 columns are numeric  and the one remaining is an object.  

In [ ]:
# Converting date column column to date format
df['Datetime'] = pd.to_datetime(df['Datetime'])

## Feature Statistics Summary

Below table provides the statistical details for each column.

In [ ]:
# Analytical summary of the data set
df.describe()

CHP_Electricity_Day(kWh)  CHP_Electricity_Night(kWh)  \
count               1488.000000                 1488.000000   
mean                 364.798387                  135.013441   
std                  274.689455                  227.427649   
min                    0.000000                    0.000000   
25%                    0.000000                    0.000000   
50%                  550.000000                    0.000000   
75%                  580.000000                  460.000000   
max                  920.000000                  900.000000   

       Boiler_1_Heat(kWh)  Boiler_2_Heat(kWh)  Boiler_3_Heat(kWh)  \
count         1488.000000         1488.000000         1488.000000   
mean             0.134409          261.088710           19.556452   
std              3.664945          203.990183           70.928643   
min              0.000000            0.000000            0.000000   
25%              0.000000          100.000000            0.000000   
50%              0.000000          200.000000            0.000000   
75%              0.000000          300.000000            0.000000   
max            100.000000         1400.000000          500.000000   

       CHP_Total_Energy(kWh)  Boiler_1_Volume(Sm3)  Boiler_2_Volume(Sm3)  \
count            1488.000000           1488.000000           1488.000000   
mean              449.798387              0.076613             31.699597   
std               160.026770              1.355272             22.433114   
min                 0.000000              0.000000              0.000000   
25%               400.000000              0.000000             19.000000   
50%               500.000000              0.000000             30.000000   
75%               500.000000              0.000000             39.000000   
max               800.000000             34.000000            172.000000   

       Boiler_3_Volume(Sm3)  CHP_2G_Volume(Sm3)  
count           1488.000000         1488.000000  
mean               2.526210          121.887769  
std                9.000746           41.230504  
min                0.000000            0.000000  
25%                0.000000          123.000000  
50%                0.000000          136.000000  
75%                0.000000          142.000000  
max               62.000000          224.000000

### Observations
1. We can see that within the 25% quartile, most of the variables have zero values, apart from *Boiler_2_Heat, CHP_Total_Energy*  and their respective volumes.
2. The standard deviations of *CHP_Electricity_Day(kWh), CHP_Electricity_Night(kWh)* and *Boiler_2_Heat(kWh)* depicts a similar behaviour, on the other hand, the mean values present a significant variation. 


## Exploratory Data Analysis

In [ ]:
# Finding missing values in the data set
df.isnull().sum()

Datetime                      0
CHP_Electricity_Day(kWh)      0
CHP_Electricity_Night(kWh)    0
Boiler_1_Heat(kWh)            0
Boiler_2_Heat(kWh)            0
Boiler_3_Heat(kWh)            0
CHP_Total_Energy(kWh)         0
Boiler_1_Volume(Sm3)          0
Boiler_2_Volume(Sm3)          0
Boiler_3_Volume(Sm3)          0
CHP_2G_Volume(Sm3)            0
dtype: int64

In [ ]:
#Finding rows containing duplicate data
duplicate_rows_df = df[df.duplicated()]
print(f'There are {0} duplicated rows in the dataset'.format(duplicate_rows_df))

There are 0 duplicated rows in the dataset


The data set does not contain either missing  nor duplicated data.

### Data Distribution Visualization - Part 1

In [ ]:
# Plotting a histogram matrix for all the variables
fig = ff.create_scatterplotmatrix(df.drop(columns= 'Datetime'), diag='histogram', height=3000, width=2600)
fig.show()

### Observations:
1. CHP Electricity generated during the day has 2 significant peaks: 0 kWh at q frequency of 500. The 600 kWh of electricity generated shows a frequency of over 700. While CHP Electricity generated during the night displays a peak at 0 kWh twice the count of the CHP used for the day.

2. In Boiler 1 all data is distributed at the mark of 0 Energy Total (kWh) 
Boiler 2 depicts a right-skewed distribution, a significant portion of the data is within the range of 0 to 400kMWh over 350 points. Reaching almost a frequency of 500. Also, the frequency decreased abruptly to a 0 value from 400 to 1400 kWh. Boiler 3 shows a similar response as Boiler 1 but presents a slight difference, it delivers less than 100 counts at 0.15 MWh, and within 200 - 300k KWh.

3. Boiler 1 gas corrected volume is grouped at  $ 0\ Sm^{3}$ which is congruent with the Heat Energy plot. Boiler 2 gas volume has its peak around $20-30\ Sm^{3}$. Boiler 3 has an expected gas volume of about 0 and a few data within 20 and 50 $Sm^{3}$. The CHP gas volume is more likely to fall roughly 100 up to 160 $Sm^{3}$. And in comparison, the Boilers' volume is notably larger.

#### Note 
$Sm^{3}$ means the volume of Gas that occupies one cubic meter $(1\ m^{3})$ at Standard Conditions

### Data Distribution Visualization - Part 2

In [ ]:
# Boxplots to get an idea of the distribution/outliers
fig = px.box(df.iloc[:,1:-1])
fig.show()

In [ ]:
# Finding the relationship between the variables
fig = px.imshow(df.corr(), text_auto=True, aspect="auto", color_continuous_scale=px.colors.sequential.Blues, height=800, width=1200)
fig.show()

### Observations: 

1. Columns *CHP_Electricity_Day(kWh)* and *CHP_Electricity_Night(kWh)* are a complementary to each other. The readings with zero value in column Day mean the CHP is generating electricity at night and the same the other way around.

2. Considerable amount of outliers in Boiler_2_Heat(kWh), Boiler_3_Heat(kWh) and their volumes.

Graph 2: 
- A clear postivie linear relationship between boilers heat and volume.
- CHP total energy relationship with volumen is a linear as well. 


3. CHP Electricity during day has a larger linear relationship with the heat energy generated than at night.  

**Question:**
Are the CHP and boilers comsumption at the same intervals of time? 

## Feature Engineering


To visualise the data by time steps values are going to be group by mean per day intervals.

In [ ]:
df = df.set_index('Datetime').groupby(pd.Grouper(freq='D')).mean()
df= df.dropna().reset_index()

In [ ]:
CHP_Electricity_df = df.iloc[:,0:3]
fig = px.line(df, x='Datetime', y=CHP_Electricity_df.columns)
fig.update_layout(
    title='CHP Electricity Day/Night',
    xaxis_title="Date time",
    yaxis_title="Electricity kWh",
    legend_title="Variables"
    )
fig.show()

In [ ]:
fig = px.line(df, x='Datetime', y='CHP_Total_Energy(kWh)')
fig.update_layout(
    title='CHP Total Heat Energy',
    xaxis_title="Date time",
    yaxis_title="Heat Energy kWh"
    )
fig.show()

- The plots above show a constant generation from 10th April to 10th August 2020 about the 500 kWh mark in Heat Energy. Additionally CHP electricity during Day performs about 400kWh and during the night roughly 150 kWh.
Subsequently, CHP electricity and heat indicator faced a fall, dropping down the 100kWh mark on Oct 10th. 
- Max peak was reached Oct 14th es remains stable till Oct 31st. 
- Another dropping point is detected on 10th Nov and continues rising from there.


In [ ]:
Boilers_df = df.iloc[:,3:6]
fig = px.line(df, x="Datetime", y=Boilers_df.columns)
fig.update_layout(
    title='Boilers Comparative',
    xaxis_title="Date time",
    yaxis_title="Heat Energy kWh",
    legend_title="Variables"
    )
fig.show()

- In contrast, Boiler usage's highest points are the same dates in which
CHP Electricity and Heat Energy experience a reduction in generation. It demostrates a inverse relationship.
- Furthermore, in the period with constant CHP Electricity and Heat generation, all boilers usage is minimal, and only Boiler_2 consumption is active at a minor level a cause may be that it compensates for the reduction of CHP Heat energy production.

## Basic Modelling 

In [ ]:
import plotly.graph_objects as go
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

X = df['CHP_Total_Energy(kWh)'].values.reshape(-1, 1)
X_train, X_test, y_train, y_test = train_test_split(X, df['Boiler_2_Heat(kWh)'], random_state=0)

model = LinearRegression()
model.fit(X_train, y_train)

x_range = np.linspace(X.min(), X.max(), 100)
y_range = model.predict(x_range.reshape(-1, 1))


fig = go.Figure([
    go.Scatter(x=X_train.squeeze(), y=y_train, name='train', mode='markers'),
    go.Scatter(x=X_test.squeeze(), y=y_test, name='test', mode='markers'),
    go.Scatter(x=x_range, y=y_range, name='prediction')
])
fig.show()

The linear regression roughly describes that the Boiler and Heat Energy 
have a relationship with an inverse variation.

## Conclusions
Both On-site Electricity and heat energy individually produce constantly over 450 kWh from April to the End of August (Spring) and that decreases through September. Reaching the lowest peak on Oct 10th, on this exact date Boiler 2 highest peak is close to 700 kWh. Although by Oct 14th the Electricity generation increases up to 500kWh for a month, on Nov 10th decreases again. Then gradually starts increasing again during Winter.

It seems that in the scenarios where CHP Heat and electricity generation declined a Boiler is used to counteract the lack of heat energy and electricity production. Even though, in this case, only Boiler 2 has been used significantly, Boiler 3 was used on a particular occasion when Boiler 2 lowered to 0 kWh on Oct 30th.